### PR TITLE
Avoid emitting NOTICE on implicit function creation/drop during ALTER TABLE

### PR DESCRIPTION
**Pull request by danielgustafsson**
When a table which had an attribute whose type has been dropped process the ALTER TABLE command queue, a "hidden" type will be created, and immediately dropped, during ALTER TABLE processing. This will emit several NOTICEs which can be confusing to the user as it's an autogenerated name and the DROP TYPE can have happened at a previous time. Below is an example of the output:

```SQL
create table <tablename> (a integer, b <typename>);
drop type <typename>;
  ...
alter table <tablename> set with(reorganize = true) distributed randomly;
NOTICE:  return type pg_atsdb_<oid>_2_3 is only a shell
NOTICE:  argument type pg_atsdb_<oid>_2_3 is only a shell
NOTICE:  drop cascades to function pg_atsdb_<oid>_2_3_out(pg_atsdb_<oid>_2_3)
NOTICE:  drop cascades to function pg_atsdb_<oid>_2_3_in(cstring)
```
In order to override the messagelevel and mute the emission of these NOTICEs, extend CREATE FUNCTION and set INTERNAL depedency on the functions based on the type.

By passing in quiet as a new WITH parameter CreateFunction() is instructed to not emit NOTICE messages for shell types.

```SQL
CREATE FUNCTION f .. WITH (quiet);
```
The in and out functions for the type will be marked as INTERNAL dependency which will instruct the backend to drop them during CASCADE without emitting NOTICEs. INTERNAL dependency implies that the objects are considered an internal implementation detail of an object and since these functions are useless without the type (and to be frank, useless with the type as well given the nature of the type) they can be considered as such.

Also remove output from expected testfiles and perform minor cleanups.

### CODE REVIEW

**Comment by danielgustafsson** [comment]
This came out of a rabbithole type comment by @hlinnaka in #1227 and while not particularly critical to fix, it was a fun hack to make so..

**Comment by hlinnaka** [comment]
Ugh, didn't realize it would be this invasive :-(. Let's step back a bit...

pg_upgrade has a similar problem. It needs to create table in the new cluster, that matches the physical layout of an existing table, including dropped columns. The way it was solved there, is that it substitutes INTEGER for all dropped columns in the CREATE TABLE commands, but and then issues a direct "UPDATE pg_attribute" command to mark the column as dropped, with the correct alignment and length. 

Perhaps we should get rid of the whole temporary type creation, and do something like that instead? A small complication is that we'd need to also dispatch the UPDATE to the segments.

One straightforward solution would be to add "attalign" and "attlen" fields to ColumnDef, and just set them directly for the dropped columns.

**Comment by hlinnaka** [comment]
BTW, any idea what this, in prebuild_temp_table(), is for?
```
				/*
				 * If this is a built in array type, like _int4, then reduce
				 * the array dimensions by 1. This is an annoying postgres
				 * hack which I wish would go away.
				 */
				if (typstr && typstr[0] == '_' && ndims > 0)
					ndims--;
```

**Comment by hlinnaka** [comment]
Actually, why do we need to reconsturct the dropped columns here at all? This code is used for ALTER TABLE SET DISTRIBUTED BY, which has to read through all data, re-distribute it, and write it again. If we're rewriting all the data, anyway, shouldn't we squeeze away all dropped columns, like other table-rewriting ALTER TABLE operations do?

**Comment by danielgustafsson** [comment]
> On 09 Dec 2016, at 17:07, Heikki Linnakangas <notifications@github.com> wrote:
> 
> BTW, any idea what this, in prebuild_temp_table(), is for?
> 
> 				/*
> 				 * If this is a built in array type, like _int4, then reduce
> 				 * the array dimensions by 1. This is an annoying postgres
> 				 * hack which I wish would go away.
> 				 */
> 				if (typstr && typstr[0] == '_' && ndims > 0)
> 					ndims—;

Without looking at the code I would guess that it accounts for the array dimensions being 1-based rather than zero-based? If I’m right, then the comment is wrong (it’s not limited to build in array types and the ‘_’ prefix is no longer guaranteed since the 8.3 merge).

**Comment by danielgustafsson** [comment]
So I took a stab at rewriting the CTAS the way you proposed and the resulting diff is so simple I assume something is wrong.. it works fine in testing though, let's see what Concourse thinks.

**Comment by danielgustafsson** [comment]
As expected things broke left/right/center in the pipeline so I'll have a look at that and will close this to re-open when I have it sorted.

**Comment by hlinnaka** [comment]
On 17 December 2016 00:20:53 EET, Daniel Gustafsson <notifications@github.com> wrote:
>So I took a stab at rewriting the CTAS the way you proposed and the
>resulting diff is so simple I assume something is wrong.. it works fine
>in testing though, let's see what Concourse thinks.

Oh wow, too easy! :-) Looks good to me at a quick glance. Great simplification if we can indeed get away with it!

It might be worth digging into the historical git repo to see when the current code was added, in case there are pointers to something we're missing. Ideally, whatever issues lead to this were memorized in regression tests, but better safe than sorry.

- Heikki

**Comment by hlinnaka** [comment]
I rebased this patch, and I attempted to fix the regression tests. Let's see what the CI thinks of it. My fix is in a separate commit in this PR, but it ought to be squashed before pushing to master. Please have a look.

File: src/backend/commands/tablecmds.c [hunk 1]
```diff
@@ -11608,16 +11349,18 @@ prebuild_temp_table(Relation rel, RangeVar *tmpname, List *distro, List *opts,
...
-				cd->colname = pstrdup(NameStr(alias));
-				drop_atts = lappend_int(drop_atts, attno);
-
-				if (built)
-					*hidden_types = lappend(*hidden_types, tname->names);
+				/*
+				 * Use dummy int4 columns to stand in for dropped columns.
+				 * We cannot easily reconstruct the original layout, because
+				 * we don't know what the original datatype was, and it might
+				 * not even exist anymoree. This means that the temp table is
```
Comment (danielgustafsson) [inline src/backend/commands/tablecmds.c]: s/anymore/anymoree/

**Comment by danielgustafsson** [comment]
From the discussions we had last week, this makes perfect sense. The pipeline seems happy as well. :+1:

**Comment by hlinnaka** [comment]
Thanks. Want to do the honors and squash and push this, or should I?

**Comment by danielgustafsson** [comment]
> On 24 Jan 2017, at 06:57, Heikki Linnakangas <notifications@github.com> wrote:
> 
> Thanks. Want to do the honors and squash and push this, or should I?

Feel free to push it in.

**Comment by hlinnaka** [comment]
Ok, squashed and pushed!

**Comment by danielgustafsson** [comment]
Thanks!

**Pull request closed without merging**